service/history/plugins/callbacks/tasks.go

```diff
@@ -0,0 +1,114 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package callbacks
+
+import (
+	"fmt"
+	"time"
+
+	"go.temporal.io/server/service/history/statemachines"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+var (
+	TaskTypeInvocation = statemachines.TaskType{
+		ID:   1,
+		Name: "callbacks.Invocation",
+	}
+	TaskTypeBackoff = statemachines.TaskType{
+		ID:   2,
+		Name: "callbacks.Backoff",
+	}
+)
+
+type InvocationTask struct {
+	Destination string
+}
+
+var _ statemachines.Task = InvocationTask{}
+
+func (InvocationTask) Type() statemachines.TaskType {
+	return TaskTypeInvocation
+}
+
+func (t InvocationTask) Category() tasks.Category {
+	return tasks.CategoryCallback
+}
+
+func (t InvocationTask) Kind() statemachines.TaskKind {
+	return statemachines.TaskKindOutbound{Destination: t.Destination}
+}
+
+type InvocationTaskSerializer struct{}
+
+func (InvocationTaskSerializer) Deserialize(data []byte, category tasks.Category, kind statemachines.TaskKind) (statemachines.Task, error) {
+	if kind, ok := kind.(statemachines.TaskKindOutbound); ok {
+		return InvocationTask{Destination: kind.Destination}, nil
+	}
+	return nil, fmt.Errorf("%w: expected outbound", statemachines.ErrInvalidTaskKind)
+}
+
+func (InvocationTaskSerializer) Serialize(statemachines.Task) ([]byte, error) {
+	return nil, nil
+}
+
+type BackoffTask struct {
+	Deadline time.Time
+}
+
+var _ statemachines.Task = BackoffTask{}
+
+func (BackoffTask) Type() statemachines.TaskType {
+	return TaskTypeBackoff
+}
+
+func (t BackoffTask) Category() tasks.Category {
+	return tasks.CategoryCallback
+}
+
+func (t BackoffTask) Kind() statemachines.TaskKind {
+	return statemachines.TaskKindTimer{Deadline: t.Deadline}
+}
+
+type BackoffTaskSerializer struct{}
+
+func (BackoffTaskSerializer) Deserialize(data []byte, category tasks.Category, kind statemachines.TaskKind) (statemachines.Task, error) {
+	if kind, ok := kind.(statemachines.TaskKindTimer); ok {
+		return BackoffTask{Deadline: kind.Deadline}, nil
+	}
+	return nil, fmt.Errorf("%w: expected timer", statemachines.ErrInvalidTaskKind)
+}
+
+func (BackoffTaskSerializer) Serialize(statemachines.Task) ([]byte, error) {
+	return nil, nil
+}
+
+func RegisterTaskSerializer(reg *statemachines.Registry) error {
```

Comment:
I'm a little curious why registering a state machine, executor, and task serializer are all separate steps. Is it possible to have one without the others?

Reply:
Yeah, I considered adding a `RegisterPlugin` method and having a more "well defined" concept but for now I'd rather have the flexibility before solidifying that interface.